jmh-benchmarks/src/main/java/org/apache/kafka/jmh/acl/AclAuthorizerBenchmark.java

```diff
@@ -69,33 +70,39 @@
 @BenchmarkMode(Mode.AverageTime)
 @OutputTimeUnit(TimeUnit.MILLISECONDS)
 public class AclAuthorizerBenchmark {
-    @Param({"10000", "50000", "200000"})
+    @Param({"10000", "40000", "80000"})
```

Comment:
I am not sure why we would make this change. If we need the change because we have become slower, we need to understand why.

Reply:
The underlying algorithm of AuthorizeByResourceType() implementation in AclAuthorizer has several characteristics:
1. If any "allow resource" of the given ACE does not have a dominant "deny resource", the API will return immediately
2. The complexity is O(n*m) where `n` is the number of "allow resources" of the given ACE, 'm' is the number of "deny resources" of the given ACE, but not related to the number of "ACE".

1) means that, given an ACE,  suppose `p%` of its "allow resource" does not have a dominant "deny resource", if `resourceCount` is `r`, on average, after checking `r * p * 0.01` "allow resources", the API will return. 
a) if we are hard-coding, like use the loop index % something to determine which "allow resource" should have a dominant "deny resource", we end up iterating the same amount of the "allow resource" and returning from the API call every time
b) if we are determine which "allow resource" should have a dominant "deny resource", the result will be too noisy.

2) means that, the API time cost is not related to the number of "ACE" but is hyperbolically increasing when `resourceCount` is increasing. Under the assumption in (1), the actual complexity would be (r * r * p * 0.01)


So what I was doing is to directly test the worst case, where p = 1, which means 100% of the "allow resource" will have a dominant "deny resource. The complexity hence would be (r^2). It's rare that a cluster can have 200k "allow resources" rules for each user and it's not fair to have a relatively smaller `aclCount` and huger `resourceCount`, as the API is indexing on `ACE`.

Reply:
I just realized that my AclAuthorizer implementation is calling `String::startWith` which also has an O(d) complexity where `d` is the length of the "deny pattern" string of the given ACL. So the complexity would be O(n * m * d). This means that iterating through all the prefixes of the `allow pattern` and check if any prefix is contained in the `deny pattern set` will be much faster, with a complexity of O(n * a), where `a` is the length of the "allow pattern" string, which should be very close to `d`. So we can say that O(n * a) < O(n * m * d). (this is what the interface default do). I'll just change the AclAuthorizer to use the same mechanism as the interface default.